{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "bi_ngs_hw4.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "67162be1b6d24da2896c004761cfd9df": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "state": {
            "_view_name": "HBoxView",
            "_dom_classes": [],
            "_model_name": "HBoxModel",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "box_style": "",
            "layout": "IPY_MODEL_d95799b233f14888b95194b5050a65c7",
            "_model_module": "@jupyter-widgets/controls",
            "children": [
              "IPY_MODEL_8732aaf4719b427bb5df4b605b94e4d2",
              "IPY_MODEL_9a505db5255541e89adb6a9fe45ba33a"
            ]
          }
        },
        "d95799b233f14888b95194b5050a65c7": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "8732aaf4719b427bb5df4b605b94e4d2": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "state": {
            "_view_name": "ProgressView",
            "style": "IPY_MODEL_49cce6045ff4464abb7740d76cf4978c",
            "_dom_classes": [],
            "description": "100%",
            "_model_name": "FloatProgressModel",
            "bar_style": "success",
            "max": 1381551,
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": 1381551,
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "orientation": "horizontal",
            "min": 0,
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_b85281b697e24408ae6959659ddb23c3"
          }
        },
        "9a505db5255541e89adb6a9fe45ba33a": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "state": {
            "_view_name": "HTMLView",
            "style": "IPY_MODEL_397768dd75d7491a8d26f4bbfc81814c",
            "_dom_classes": [],
            "description": "",
            "_model_name": "HTMLModel",
            "placeholder": "​",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": " 1381551/1381551 [03:29&lt;00:00, 6602.38it/s]",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_5ebda31e1ef8457da4ed944a0d2bfdf9"
          }
        },
        "49cce6045ff4464abb7740d76cf4978c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "ProgressStyleModel",
            "description_width": "initial",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "bar_color": null,
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "b85281b697e24408ae6959659ddb23c3": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "397768dd75d7491a8d26f4bbfc81814c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "DescriptionStyleModel",
            "description_width": "",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "5ebda31e1ef8457da4ed944a0d2bfdf9": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "4c145b72aa274c14aab271fd7bec0e5c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "state": {
            "_view_name": "HBoxView",
            "_dom_classes": [],
            "_model_name": "HBoxModel",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "box_style": "",
            "layout": "IPY_MODEL_4e4dc3b55948491b91df547fb41269ed",
            "_model_module": "@jupyter-widgets/controls",
            "children": [
              "IPY_MODEL_ba4dfd53d44f40f3854a6efa04e93515",
              "IPY_MODEL_966f7a5c714e409bb9f5a518e24a486c"
            ]
          }
        },
        "4e4dc3b55948491b91df547fb41269ed": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "ba4dfd53d44f40f3854a6efa04e93515": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "state": {
            "_view_name": "ProgressView",
            "style": "IPY_MODEL_d7536ba565414217b20ca64b122f8f7b",
            "_dom_classes": [],
            "description": "100%",
            "_model_name": "FloatProgressModel",
            "bar_style": "success",
            "max": 1381551,
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": 1381551,
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "orientation": "horizontal",
            "min": 0,
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_3ff6df99cf824775b9b6ba3c614f41ea"
          }
        },
        "966f7a5c714e409bb9f5a518e24a486c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "state": {
            "_view_name": "HTMLView",
            "style": "IPY_MODEL_5dd9329e4a3f45aa9c7998b85a88b884",
            "_dom_classes": [],
            "description": "",
            "_model_name": "HTMLModel",
            "placeholder": "​",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": " 1381551/1381551 [03:11&lt;00:00, 7228.31it/s]",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_87bfba00b1b94ff2ae133a2f2bbb23f3"
          }
        },
        "d7536ba565414217b20ca64b122f8f7b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "ProgressStyleModel",
            "description_width": "initial",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "bar_color": null,
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "3ff6df99cf824775b9b6ba3c614f41ea": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "5dd9329e4a3f45aa9c7998b85a88b884": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "DescriptionStyleModel",
            "description_width": "",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "87bfba00b1b94ff2ae133a2f2bbb23f3": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        }
      }
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qXytNiZ_AJ6t"
      },
      "source": [
        "# Insalling packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RoXsjCkivFPH"
      },
      "source": [
        "!sudo apt install trimmomatic"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hgoT8Cbt-zZg",
        "outputId": "29bca17c-4b74-4919-b60d-60c59d48ccd4"
      },
      "source": [
        "!sudo apt-get install bwa\n",
        "!sudo apt-get install samtools"
      ],
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "Suggested packages:\n",
            "  samtools\n",
            "The following NEW packages will be installed:\n",
            "  bwa\n",
            "0 upgraded, 1 newly installed, 0 to remove and 30 not upgraded.\n",
            "Need to get 176 kB of archives.\n",
            "After this operation, 404 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 bwa amd64 0.7.17-1ubuntu0.1 [176 kB]\n",
            "Fetched 176 kB in 1s (342 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 1.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package bwa.\n",
            "(Reading database ... 162485 files and directories currently installed.)\n",
            "Preparing to unpack .../bwa_0.7.17-1ubuntu0.1_amd64.deb ...\n",
            "Unpacking bwa (0.7.17-1ubuntu0.1) ...\n",
            "Setting up bwa (0.7.17-1ubuntu0.1) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  cwltool libc-ares2 libhts2 libjs-bootstrap nodejs nodejs-doc\n",
            "  python-asn1crypto python-avro python-cachecontrol python-certifi\n",
            "  python-cffi-backend python-chardet python-cryptography python-enum34\n",
            "  python-html5lib python-idna python-ipaddress python-isodate python-lockfile\n",
            "  python-mistune python-openssl python-pkg-resources python-pyparsing\n",
            "  python-rdflib python-rdflib-jsonld python-requests python-ruamel.yaml\n",
            "  python-schema-salad python-shellescape python-six python-sparqlwrapper\n",
            "  python-typing python-urllib3 python-webencodings\n",
            "Suggested packages:\n",
            "  python-cryptography-doc python-cryptography-vectors python-enum34-doc\n",
            "  python-genshi python-lxml python-lockfile-doc python-openssl-doc\n",
            "  python-openssl-dbg python-setuptools python-pyparsing-doc python-rdflib-doc\n",
            "  python-rdflib-tools python-socks python-ntlm\n",
            "The following NEW packages will be installed:\n",
            "  cwltool libc-ares2 libhts2 libjs-bootstrap nodejs nodejs-doc\n",
            "  python-asn1crypto python-avro python-cachecontrol python-certifi\n",
            "  python-cffi-backend python-chardet python-cryptography python-enum34\n",
            "  python-html5lib python-idna python-ipaddress python-isodate python-lockfile\n",
            "  python-mistune python-openssl python-pkg-resources python-pyparsing\n",
            "  python-rdflib python-rdflib-jsonld python-requests python-ruamel.yaml\n",
            "  python-schema-salad python-shellescape python-six python-sparqlwrapper\n",
            "  python-typing python-urllib3 python-webencodings samtools\n",
            "0 upgraded, 35 newly installed, 0 to remove and 30 not upgraded.\n",
            "Need to get 8,568 kB of archives.\n",
            "After this operation, 41.0 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-pkg-resources all 39.0.1-2 [128 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-six all 1.11.0-2 [11.3 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-isodate all 0.6.0-1 [23.8 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-pyparsing all 2.2.0+dfsg1-2 [52.1 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-rdflib all 4.2.1-2 [237 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-certifi all 2018.1.18-2 [144 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-chardet all 3.0.4-1 [80.3 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-idna all 2.6-1 [32.4 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 python-urllib3 all 1.22-1ubuntu0.18.04.2 [86.0 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 python-requests all 2.18.4-2ubuntu0.1 [58.5 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-ruamel.yaml amd64 0.15.34-1 [183 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-cachecontrol all 0.11.7-1 [15.2 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-lockfile all 1:0.12.2-2 [14.6 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-mistune all 0.8.3-2 [13.8 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-rdflib-jsonld all 0.4.0-3 [12.2 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-typing all 3.6.2-1 [22.2 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-avro all 1.8.2+dfsg-1 [23.2 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libjs-bootstrap all 3.3.7+dfsg-2 [228 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-schema-salad all 2.6.20171201034858-3 [260 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-shellescape all 3.4.1-2 [4,432 B]\n",
            "Get:21 http://archive.ubuntu.com/ubuntu bionic/universe amd64 cwltool all 1.0.20180302231433-1 [143 kB]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libhts2 amd64 1.7-2 [300 kB]\n",
            "Get:23 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 nodejs-doc all 8.10.0~dfsg-2ubuntu0.4 [752 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-asn1crypto all 0.24.0-1 [72.7 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-cffi-backend amd64 1.11.5-1 [63.4 kB]\n",
            "Get:26 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-enum34 all 1.1.6-2 [34.8 kB]\n",
            "Get:27 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-ipaddress all 1.0.17-1 [18.2 kB]\n",
            "Get:28 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 python-cryptography amd64 2.1.4-1ubuntu1.4 [276 kB]\n",
            "Get:29 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-webencodings all 0.5-2 [10.3 kB]\n",
            "Get:30 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-html5lib all 0.999999999-1 [83.6 kB]\n",
            "Get:31 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-openssl all 17.5.0-1ubuntu1 [41.3 kB]\n",
            "Get:32 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-sparqlwrapper all 1.7.6-2 [21.8 kB]\n",
            "Get:33 http://archive.ubuntu.com/ubuntu bionic/universe amd64 samtools amd64 1.7-1 [288 kB]\n",
            "Get:34 http://archive.ubuntu.com/ubuntu bionic/main amd64 libc-ares2 amd64 1.14.0-1 [37.1 kB]\n",
            "Get:35 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 nodejs amd64 8.10.0~dfsg-2ubuntu0.4 [4,796 kB]\n",
            "Fetched 8,568 kB in 1s (6,344 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 35.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package python-pkg-resources.\n",
            "(Reading database ... 162502 files and directories currently installed.)\n",
            "Preparing to unpack .../00-python-pkg-resources_39.0.1-2_all.deb ...\n",
            "Unpacking python-pkg-resources (39.0.1-2) ...\n",
            "Selecting previously unselected package python-six.\n",
            "Preparing to unpack .../01-python-six_1.11.0-2_all.deb ...\n",
            "Unpacking python-six (1.11.0-2) ...\n",
            "Selecting previously unselected package python-isodate.\n",
            "Preparing to unpack .../02-python-isodate_0.6.0-1_all.deb ...\n",
            "Unpacking python-isodate (0.6.0-1) ...\n",
            "Selecting previously unselected package python-pyparsing.\n",
            "Preparing to unpack .../03-python-pyparsing_2.2.0+dfsg1-2_all.deb ...\n",
            "Unpacking python-pyparsing (2.2.0+dfsg1-2) ...\n",
            "Selecting previously unselected package python-rdflib.\n",
            "Preparing to unpack .../04-python-rdflib_4.2.1-2_all.deb ...\n",
            "Unpacking python-rdflib (4.2.1-2) ...\n",
            "Selecting previously unselected package python-certifi.\n",
            "Preparing to unpack .../05-python-certifi_2018.1.18-2_all.deb ...\n",
            "Unpacking python-certifi (2018.1.18-2) ...\n",
            "Selecting previously unselected package python-chardet.\n",
            "Preparing to unpack .../06-python-chardet_3.0.4-1_all.deb ...\n",
            "Unpacking python-chardet (3.0.4-1) ...\n",
            "Selecting previously unselected package python-idna.\n",
            "Preparing to unpack .../07-python-idna_2.6-1_all.deb ...\n",
            "Unpacking python-idna (2.6-1) ...\n",
            "Selecting previously unselected package python-urllib3.\n",
            "Preparing to unpack .../08-python-urllib3_1.22-1ubuntu0.18.04.2_all.deb ...\n",
            "Unpacking python-urllib3 (1.22-1ubuntu0.18.04.2) ...\n",
            "Selecting previously unselected package python-requests.\n",
            "Preparing to unpack .../09-python-requests_2.18.4-2ubuntu0.1_all.deb ...\n",
            "Unpacking python-requests (2.18.4-2ubuntu0.1) ...\n",
            "Selecting previously unselected package python-ruamel.yaml.\n",
            "Preparing to unpack .../10-python-ruamel.yaml_0.15.34-1_amd64.deb ...\n",
            "Unpacking python-ruamel.yaml (0.15.34-1) ...\n",
            "Selecting previously unselected package python-cachecontrol.\n",
            "Preparing to unpack .../11-python-cachecontrol_0.11.7-1_all.deb ...\n",
            "Unpacking python-cachecontrol (0.11.7-1) ...\n",
            "Selecting previously unselected package python-lockfile.\n",
            "Preparing to unpack .../12-python-lockfile_1%3a0.12.2-2_all.deb ...\n",
            "Unpacking python-lockfile (1:0.12.2-2) ...\n",
            "Selecting previously unselected package python-mistune.\n",
            "Preparing to unpack .../13-python-mistune_0.8.3-2_all.deb ...\n",
            "Unpacking python-mistune (0.8.3-2) ...\n",
            "Selecting previously unselected package python-rdflib-jsonld.\n",
            "Preparing to unpack .../14-python-rdflib-jsonld_0.4.0-3_all.deb ...\n",
            "Unpacking python-rdflib-jsonld (0.4.0-3) ...\n",
            "Selecting previously unselected package python-typing.\n",
            "Preparing to unpack .../15-python-typing_3.6.2-1_all.deb ...\n",
            "Unpacking python-typing (3.6.2-1) ...\n",
            "Selecting previously unselected package python-avro.\n",
            "Preparing to unpack .../16-python-avro_1.8.2+dfsg-1_all.deb ...\n",
            "Unpacking python-avro (1.8.2+dfsg-1) ...\n",
            "Selecting previously unselected package libjs-bootstrap.\n",
            "Preparing to unpack .../17-libjs-bootstrap_3.3.7+dfsg-2_all.deb ...\n",
            "Unpacking libjs-bootstrap (3.3.7+dfsg-2) ...\n",
            "Selecting previously unselected package python-schema-salad.\n",
            "Preparing to unpack .../18-python-schema-salad_2.6.20171201034858-3_all.deb ...\n",
            "Unpacking python-schema-salad (2.6.20171201034858-3) ...\n",
            "Selecting previously unselected package python-shellescape.\n",
            "Preparing to unpack .../19-python-shellescape_3.4.1-2_all.deb ...\n",
            "Unpacking python-shellescape (3.4.1-2) ...\n",
            "Selecting previously unselected package cwltool.\n",
            "Preparing to unpack .../20-cwltool_1.0.20180302231433-1_all.deb ...\n",
            "Unpacking cwltool (1.0.20180302231433-1) ...\n",
            "Selecting previously unselected package libhts2:amd64.\n",
            "Preparing to unpack .../21-libhts2_1.7-2_amd64.deb ...\n",
            "Unpacking libhts2:amd64 (1.7-2) ...\n",
            "Selecting previously unselected package nodejs-doc.\n",
            "Preparing to unpack .../22-nodejs-doc_8.10.0~dfsg-2ubuntu0.4_all.deb ...\n",
            "Unpacking nodejs-doc (8.10.0~dfsg-2ubuntu0.4) ...\n",
            "Selecting previously unselected package python-asn1crypto.\n",
            "Preparing to unpack .../23-python-asn1crypto_0.24.0-1_all.deb ...\n",
            "Unpacking python-asn1crypto (0.24.0-1) ...\n",
            "Selecting previously unselected package python-cffi-backend.\n",
            "Preparing to unpack .../24-python-cffi-backend_1.11.5-1_amd64.deb ...\n",
            "Unpacking python-cffi-backend (1.11.5-1) ...\n",
            "Selecting previously unselected package python-enum34.\n",
            "Preparing to unpack .../25-python-enum34_1.1.6-2_all.deb ...\n",
            "Unpacking python-enum34 (1.1.6-2) ...\n",
            "Selecting previously unselected package python-ipaddress.\n",
            "Preparing to unpack .../26-python-ipaddress_1.0.17-1_all.deb ...\n",
            "Unpacking python-ipaddress (1.0.17-1) ...\n",
            "Selecting previously unselected package python-cryptography.\n",
            "Preparing to unpack .../27-python-cryptography_2.1.4-1ubuntu1.4_amd64.deb ...\n",
            "Unpacking python-cryptography (2.1.4-1ubuntu1.4) ...\n",
            "Selecting previously unselected package python-webencodings.\n",
            "Preparing to unpack .../28-python-webencodings_0.5-2_all.deb ...\n",
            "Unpacking python-webencodings (0.5-2) ...\n",
            "Selecting previously unselected package python-html5lib.\n",
            "Preparing to unpack .../29-python-html5lib_0.999999999-1_all.deb ...\n",
            "Unpacking python-html5lib (0.999999999-1) ...\n",
            "Selecting previously unselected package python-openssl.\n",
            "Preparing to unpack .../30-python-openssl_17.5.0-1ubuntu1_all.deb ...\n",
            "Unpacking python-openssl (17.5.0-1ubuntu1) ...\n",
            "Selecting previously unselected package python-sparqlwrapper.\n",
            "Preparing to unpack .../31-python-sparqlwrapper_1.7.6-2_all.deb ...\n",
            "Unpacking python-sparqlwrapper (1.7.6-2) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../32-samtools_1.7-1_amd64.deb ...\n",
            "Unpacking samtools (1.7-1) ...\n",
            "Selecting previously unselected package libc-ares2:amd64.\n",
            "Preparing to unpack .../33-libc-ares2_1.14.0-1_amd64.deb ...\n",
            "Unpacking libc-ares2:amd64 (1.14.0-1) ...\n",
            "Selecting previously unselected package nodejs.\n",
            "Preparing to unpack .../34-nodejs_8.10.0~dfsg-2ubuntu0.4_amd64.deb ...\n",
            "Unpacking nodejs (8.10.0~dfsg-2ubuntu0.4) ...\n",
            "Setting up python-idna (2.6-1) ...\n",
            "Setting up nodejs-doc (8.10.0~dfsg-2ubuntu0.4) ...\n",
            "Setting up python-asn1crypto (0.24.0-1) ...\n",
            "Setting up python-mistune (0.8.3-2) ...\n",
            "Setting up python-certifi (2018.1.18-2) ...\n",
            "Setting up python-avro (1.8.2+dfsg-1) ...\n",
            "Setting up python-pkg-resources (39.0.1-2) ...\n",
            "Setting up python-ruamel.yaml (0.15.34-1) ...\n",
            "Setting up python-cffi-backend (1.11.5-1) ...\n",
            "Setting up python-six (1.11.0-2) ...\n",
            "Setting up python-isodate (0.6.0-1) ...\n",
            "Setting up python-shellescape (3.4.1-2) ...\n",
            "Setting up python-enum34 (1.1.6-2) ...\n",
            "Setting up libjs-bootstrap (3.3.7+dfsg-2) ...\n",
            "Setting up python-lockfile (1:0.12.2-2) ...\n",
            "Setting up python-ipaddress (1.0.17-1) ...\n",
            "Setting up libc-ares2:amd64 (1.14.0-1) ...\n",
            "Setting up libhts2:amd64 (1.7-2) ...\n",
            "Setting up python-webencodings (0.5-2) ...\n",
            "Setting up python-typing (3.6.2-1) ...\n",
            "Setting up python-pyparsing (2.2.0+dfsg1-2) ...\n",
            "Setting up python-urllib3 (1.22-1ubuntu0.18.04.2) ...\n",
            "Setting up python-chardet (3.0.4-1) ...\n",
            "Setting up nodejs (8.10.0~dfsg-2ubuntu0.4) ...\n",
            "update-alternatives: using /usr/bin/nodejs to provide /usr/bin/js (js) in auto mode\n",
            "Setting up python-rdflib (4.2.1-2) ...\n",
            "Setting up python-sparqlwrapper (1.7.6-2) ...\n",
            "Setting up python-rdflib-jsonld (0.4.0-3) ...\n",
            "Setting up python-cryptography (2.1.4-1ubuntu1.4) ...\n",
            "Setting up python-requests (2.18.4-2ubuntu0.1) ...\n",
            "Setting up python-cachecontrol (0.11.7-1) ...\n",
            "Setting up python-schema-salad (2.6.20171201034858-3) ...\n",
            "Setting up python-html5lib (0.999999999-1) ...\n",
            "Setting up samtools (1.7-1) ...\n",
            "Setting up python-openssl (17.5.0-1ubuntu1) ...\n",
            "Setting up cwltool (1.0.20180302231433-1) ...\n",
            "Processing triggers for libc-bin (2.27-3ubuntu1.2) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\n",
            "\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3QfLVqDpwGEH",
        "outputId": "44429f56-d35d-427f-c7b8-c6055a0360b9"
      },
      "source": [
        "!wget http://cab.spbu.ru/files/release3.10.1/SPAdes-3.10.1-Linux.tar.gz\n",
        "!tar -xzf SPAdes-3.10.1-Linux.tar.gz"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-04-04 16:54:53--  http://cab.spbu.ru/files/release3.10.1/SPAdes-3.10.1-Linux.tar.gz\n",
            "Resolving cab.spbu.ru (cab.spbu.ru)... 195.70.219.98\n",
            "Connecting to cab.spbu.ru (cab.spbu.ru)|195.70.219.98|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 10294085 (9.8M) [application/octet-stream]\n",
            "Saving to: ‘SPAdes-3.10.1-Linux.tar.gz’\n",
            "\n",
            "SPAdes-3.10.1-Linux 100%[===================>]   9.82M  7.19MB/s    in 1.4s    \n",
            "\n",
            "2021-04-04 16:54:54 (7.19 MB/s) - ‘SPAdes-3.10.1-Linux.tar.gz’ saved [10294085/10294085]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KZmMv3jjAWNK"
      },
      "source": [
        "# Aligning raw reads"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tkBf15BqAvgg",
        "outputId": "c4334d46-97f7-4af3-a5ec-cf582d458b98"
      },
      "source": [
        "!bwa index MG1655-K12.first400K.fasta\n",
        "!bwa mem MG1655-K12.first400K.fasta ecoli_400K_err_1.fastq.gz ecoli_400K_err_2.fastq.gz > alignment.sam\n",
        "!samtools view -S -b alignment.sam > alignment.bam\n",
        "!samtools sort alignment.bam > alignment_sorted.bam\n",
        "!samtools index alignment_sorted.bam"
      ],
      "execution_count": 37,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[bwa_index] Pack FASTA... 0.01 sec\n",
            "[bwa_index] Construct BWT for the packed sequence...\n",
            "[bwa_index] 0.09 seconds elapse.\n",
            "[bwa_index] Update BWT... 0.00 sec\n",
            "[bwa_index] Pack forward-only FASTA... 0.00 sec\n",
            "[bwa_index] Construct SA from BWT and Occ... 0.04 sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa index MG1655-K12.first400K.fasta\n",
            "[main] Real time: 0.227 sec; CPU: 0.196 sec\n",
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48301, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.27, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.031 CPU sec, 3.725 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48357, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.22, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.028 CPU sec, 3.633 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48345, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.039 CPU sec, 3.665 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48337, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.37, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.046 CPU sec, 3.619 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48338, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.35, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.106 CPU sec, 3.854 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48367, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.33, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.108 CPU sec, 3.698 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48287, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
            "[M::mem_pestat] mean and std.dev: (214.23, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.255 CPU sec, 3.853 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48328, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.36, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.141 CPU sec, 3.747 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48359, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.37, 10.08)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.212 CPU sec, 3.835 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48323, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.38, 10.12)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.250 CPU sec, 3.807 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48276, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.26, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.217 CPU sec, 3.836 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48231, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.27, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.260 CPU sec, 3.857 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48187, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.35, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.438 CPU sec, 4.057 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48265, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.41, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.216 CPU sec, 3.795 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48287, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.041 CPU sec, 3.634 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48307, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.33, 10.12)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.045 CPU sec, 3.661 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48303, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.40, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.035 CPU sec, 3.642 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48393, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.40, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.928 CPU sec, 3.586 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48320, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.32, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.897 CPU sec, 3.507 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48377, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.26, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.867 CPU sec, 3.438 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48307, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.31, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.931 CPU sec, 3.532 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48322, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
            "[M::mem_pestat] mean and std.dev: (214.27, 10.06)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 4.021 CPU sec, 3.605 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48361, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.23, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.841 CPU sec, 3.457 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48350, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.899 CPU sec, 3.450 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48319, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.32, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.817 CPU sec, 3.448 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48334, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.806 CPU sec, 3.505 real sec\n",
            "[M::process] read 63204 sequences (6320400 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48289, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.28, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.964 CPU sec, 3.797 real sec\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 30551, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.31, 10.19)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 63204 reads in 2.850 CPU sec, 2.649 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem MG1655-K12.first400K.fasta ecoli_400K_err_1.fastq.gz ecoli_400K_err_2.fastq.gz\n",
            "[main] Real time: 102.727 sec; CPU: 112.670 sec\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "d_juejwLG9mL",
        "outputId": "51feb5e7-78c5-47c5-c12b-f7e28da623ec"
      },
      "source": [
        "!samtools flagstat alignment_sorted.bam"
      ],
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "2763204 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "0 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "2762617 + 0 mapped (99.98% : N/A)\n",
            "2763204 + 0 paired in sequencing\n",
            "1381602 + 0 read1\n",
            "1381602 + 0 read2\n",
            "2760820 + 0 properly paired (99.91% : N/A)\n",
            "2762132 + 0 with itself and mate mapped\n",
            "485 + 0 singletons (0.02% : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3sG6NhUPAZAl"
      },
      "source": [
        "# Aligning SPADES result "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GYNNcDM1xMfX",
        "outputId": "2d15d713-070c-4265-a266-62ddf25f7c5b"
      },
      "source": [
        "!python3.5 ./SPAdes-3.10.1-Linux/bin/spades.py --only-error-correction --pe1-1 ecoli_400K_err_1.fastq.gz --pe1-2 ecoli_400K_err_2.fastq.gz -o corrected_spades"
      ],
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Command line: ./SPAdes-3.10.1-Linux/bin/spades.py\t--only-error-correction\t--pe1-1\t/content/ecoli_400K_err_1.fastq.gz\t--pe1-2\t/content/ecoli_400K_err_2.fastq.gz\t-o\t/content/corrected_spades\t\n",
            "\n",
            "System information:\n",
            "  SPAdes version: 3.10.1\n",
            "  Python version: 3.5.10\n",
            "  OS: Linux-4.19.112+-x86_64-with-Ubuntu-18.04-bionic\n",
            "\n",
            "Output dir: /content/corrected_spades\n",
            "Mode: ONLY read error correction (without assembling)\n",
            "Debug mode is turned OFF\n",
            "\n",
            "Dataset parameters:\n",
            "  Multi-cell mode (you should set '--sc' flag if input data was obtained with MDA (single-cell) technology or --meta flag if processing metagenomic dataset)\n",
            "  Reads:\n",
            "    Library number: 1, library type: paired-end\n",
            "      orientation: fr\n",
            "      left reads: ['/content/ecoli_400K_err_1.fastq.gz']\n",
            "      right reads: ['/content/ecoli_400K_err_2.fastq.gz']\n",
            "      interlaced reads: not specified\n",
            "      single reads: not specified\n",
            "Read error correction parameters:\n",
            "  Iterations: 1\n",
            "  PHRED offset will be auto-detected\n",
            "  Corrected reads will be compressed (with gzip)\n",
            "Other parameters:\n",
            "  Dir for temp files: /content/corrected_spades/tmp\n",
            "  Threads: 16\n",
            "  Memory limit (in Gb): 12\n",
            "\n",
            "\n",
            "======= SPAdes pipeline started. Log can be found here: /content/corrected_spades/spades.log\n",
            "\n",
            "\n",
            "===== Read error correction started. \n",
            "\n",
            "\n",
            "== Running read error correction tool: /content/SPAdes-3.10.1-Linux/bin/hammer /content/corrected_spades/corrected/configs/config.info\n",
            "\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  83)   Starting BayesHammer, built from N/A, git revision N/A\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  84)   Loading config from /content/corrected_spades/corrected/configs/config.info\n",
            "  0:00:00.001     4M / 4M    INFO    General                 (memory_limit.hpp          :  47)   Memory limit set to 12 Gb\n",
            "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  :  93)   Trying to determine PHRED offset\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (main.cpp                  :  99)   Determined value is 33\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (hammer_tools.cpp          :  36)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (main.cpp                  : 120)   Size of aux. kmer data 24 bytes\n",
            "     === ITERATION 0 begins ===\n",
            "  0:00:00.003     4M / 4M    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 428)   Building kmer index\n",
            "  0:00:00.003     4M / 4M    INFO   K-mer Splitting          (kmer_data.cpp             :  91)   Splitting kmer instances into 32 buckets. This might take a while.\n",
            "  0:00:00.003     4M / 4M    INFO    General                 (file_limit.hpp            :  30)   Open file limit set to 1048576\n",
            "  0:00:00.003     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 108)   Memory available for splitting buffers: 1.99935 Gb\n",
            "  0:00:00.003     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 116)   Using cell size of 2097152\n",
            "  0:00:00.003     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             :  98)   Processing /content/ecoli_400K_err_1.fastq.gz\n",
            "  0:00:13.465     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 474261 reads\n",
            "  0:00:26.630     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 947703 reads\n",
            "  0:00:39.109     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 1381602 reads\n",
            "  0:00:39.109     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             :  98)   Processing /content/ecoli_400K_err_2.fastq.gz\n",
            "  0:00:52.682     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 1865024 reads\n",
            "  0:01:06.108     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 2344539 reads\n",
            "  0:01:18.188     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 2763204 reads\n",
            "  0:01:18.188     1G / 1G    INFO   K-mer Splitting          (kmer_data.cpp             : 113)   Total 2763204 reads processed\n",
            "  0:01:18.228     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 252)   Starting k-mer counting.\n",
            "  0:01:19.632     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 258)   K-mer counting done. There are 13488952 kmers in total.\n",
            "  0:01:19.632     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 260)   Merging temporary buckets.\n",
            "  0:01:19.817     8M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 437)   Building perfect hash indices\n",
            "  0:01:22.815     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 276)   Merging final buckets.\n",
            "  0:01:22.971     8M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 483)   Index built. Total 4407256 bytes occupied (2.61385 bits per kmer).\n",
            "  0:01:22.971     8M / 1G    INFO   K-mer Counting           (kmer_data.cpp             : 359)   Arranging kmers in hash map order\n",
            "  0:01:25.022   216M / 1G    INFO    General                 (main.cpp                  : 155)   Clustering Hamming graph.\n",
            "  0:03:26.377   216M / 1G    INFO    General                 (main.cpp                  : 162)   Extracting clusters\n",
            "  0:03:36.948   216M / 1G    INFO    General                 (main.cpp                  : 174)   Clustering done. Total clusters: 2187352\n",
            "  0:03:36.953   112M / 1G    INFO   K-mer Counting           (kmer_data.cpp             : 381)   Collecting K-mer information, this takes a while.\n",
            "  0:03:37.134   424M / 1G    INFO   K-mer Counting           (kmer_data.cpp             : 387)   Processing /content/ecoli_400K_err_1.fastq.gz\n",
            "  0:07:30.578   424M / 1G    INFO   K-mer Counting           (kmer_data.cpp             : 387)   Processing /content/ecoli_400K_err_2.fastq.gz\n",
            "  0:11:32.011   424M / 1G    INFO   K-mer Counting           (kmer_data.cpp             : 394)   Collection done, postprocessing.\n",
            "  0:11:32.087   424M / 1G    INFO   K-mer Counting           (kmer_data.cpp             : 408)   There are 13488952 kmers in total. Among them 9831116 (72.8827%) are singletons.\n",
            "  0:11:32.087   424M / 1G    INFO    General                 (main.cpp                  : 180)   Subclustering Hamming graph\n",
            "  0:11:47.469   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 649)   Subclustering done. Total 1 non-read kmers were generated.\n",
            "  0:11:47.469   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 650)   Subclustering statistics:\n",
            "  0:11:47.469   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 651)     Total singleton hamming clusters: 1393328. Among them 2580 (0.185168%) are good\n",
            "  0:11:47.469   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 652)     Total singleton subclusters: 4144. Among them 4142 (99.9517%) are good\n",
            "  0:11:47.469   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 653)     Total non-singleton subcluster centers: 855730. Among them 836830 (97.7914%) are good\n",
            "  0:11:47.470   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 654)     Average size of non-trivial subcluster: 14.1349 kmers\n",
            "  0:11:47.470   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 655)     Average number of sub-clusters per non-singleton cluster: 1.08293\n",
            "  0:11:47.470   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 656)     Total solid k-mers: 843552\n",
            "  0:11:47.470   424M / 1G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 657)     Substitution probabilities: [4,4]((0.925362,0.0417852,0.0246926,0.00816041),(0.0168246,0.960387,0.0122684,0.0105199),(0.0104933,0.0122445,0.960404,0.0168578),(0.00814033,0.0246657,0.0417494,0.925445))\n",
            "  0:11:47.473   424M / 1G    INFO    General                 (main.cpp                  : 185)   Finished clustering.\n",
            "  0:11:47.473   424M / 1G    INFO    General                 (main.cpp                  : 204)   Starting solid k-mers expansion in 2 threads.\n",
            "  0:14:18.289   424M / 1G    INFO    General                 (main.cpp                  : 225)   Solid k-mers iteration 0 produced 45153 new k-mers.\n",
            "  0:16:53.084   424M / 1G    INFO    General                 (main.cpp                  : 225)   Solid k-mers iteration 1 produced 1685 new k-mers.\n",
            "  0:19:16.650   424M / 1G    INFO    General                 (main.cpp                  : 225)   Solid k-mers iteration 2 produced 63 new k-mers.\n",
            "  0:21:34.831   424M / 1G    INFO    General                 (main.cpp                  : 225)   Solid k-mers iteration 3 produced 32 new k-mers.\n",
            "  0:23:53.699   424M / 1G    INFO    General                 (main.cpp                  : 225)   Solid k-mers iteration 4 produced 0 new k-mers.\n",
            "  0:23:53.699   424M / 1G    INFO    General                 (main.cpp                  : 229)   Solid k-mers finalized\n",
            "  0:23:53.699   424M / 1G    INFO    General                 (hammer_tools.cpp          : 211)   Starting read correction in 2 threads.\n",
            "  0:23:53.699   424M / 1G    INFO    General                 (hammer_tools.cpp          : 222)   Correcting pair of reads: /content/ecoli_400K_err_1.fastq.gz and /content/ecoli_400K_err_2.fastq.gz\n",
            "  0:23:54.882   560M / 1G    INFO    General                 (hammer_tools.cpp          : 166)   Prepared batch 0 of 200000 reads.\n",
            "  0:24:05.282   580M / 1G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 0\n",
            "  0:24:05.971   580M / 1G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 0\n",
            "  0:24:07.031   576M / 1G    INFO    General                 (hammer_tools.cpp          : 166)   Prepared batch 1 of 200000 reads.\n",
            "  0:24:17.467   572M / 1G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 1\n",
            "  0:24:18.175   572M / 1G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 1\n",
            "  0:24:19.231   572M / 1G    INFO    General                 (hammer_tools.cpp          : 166)   Prepared batch 2 of 200000 reads.\n",
            "  0:24:29.336   572M / 1G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 2\n",
            "  0:24:30.090   572M / 1G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 2\n",
            "  0:24:31.129   572M / 1G    INFO    General                 (hammer_tools.cpp          : 166)   Prepared batch 3 of 200000 reads.\n",
            "  0:24:41.273   572M / 1G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 3\n",
            "  0:24:42.035   572M / 1G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 3\n",
            "  0:24:43.072   572M / 1G    INFO    General                 (hammer_tools.cpp          : 166)   Prepared batch 4 of 200000 reads.\n",
            "  0:24:53.423   572M / 1G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 4\n",
            "  0:24:54.087   572M / 1G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 4\n",
            "  0:24:55.124   572M / 1G    INFO    General                 (hammer_tools.cpp          : 166)   Prepared batch 5 of 200000 reads.\n",
            "  0:25:05.523   572M / 1G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 5\n",
            "  0:25:06.140   572M / 1G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 5\n",
            "  0:25:07.085   572M / 1G    INFO    General                 (hammer_tools.cpp          : 166)   Prepared batch 6 of 181602 reads.\n",
            "  0:25:16.647   572M / 1G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 6\n",
            "  0:25:17.150   572M / 1G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 6\n",
            "  0:25:17.305   424M / 1G    INFO    General                 (hammer_tools.cpp          : 270)   Correction done. Changed 883906 bases in 651928 reads.\n",
            "  0:25:17.305   424M / 1G    INFO    General                 (hammer_tools.cpp          : 271)   Failed to correct 21971 bases out of 252262696.\n",
            "  0:25:17.322     8M / 1G    INFO    General                 (main.cpp                  : 262)   Saving corrected dataset description to /content/corrected_spades/corrected/corrected.yaml\n",
            "  0:25:17.322     8M / 1G    INFO    General                 (main.cpp                  : 269)   All done. Exiting.\n",
            "\n",
            "== Compressing corrected reads (with gzip)\n",
            "\n",
            "== Dataset description file was created: /content/corrected_spades/corrected/corrected.yaml\n",
            "\n",
            "\n",
            "===== Read error correction finished. \n",
            "\n",
            " * Corrected reads are in /content/corrected_spades/corrected/\n",
            "\n",
            "======= SPAdes pipeline finished.\n",
            "\n",
            "SPAdes log can be found here: /content/corrected_spades/spades.log\n",
            "\n",
            "Thank you for using SPAdes!\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_MiFhYRs-FvL",
        "outputId": "bd039916-e457-4d35-a7ac-cb7a633181d9"
      },
      "source": [
        "!bwa index MG1655-K12.first400K.fasta\n",
        "!bwa mem MG1655-K12.first400K.fasta corrected_spades/corrected/ecoli_400K_err_1.fastq.00.0_0.cor.fastq.gz corrected_spades/corrected/ecoli_400K_err_2.fastq.00.0_0.cor.fastq.gz > alignment_spades.sam\n",
        "!samtools view -S -b alignment_spades.sam > alignment_spades.bam\n",
        "!samtools sort alignment_spades.bam > alignment_spades_sorted.bam\n",
        "!samtools index alignment_spades_sorted.bam"
      ],
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[bwa_index] Pack FASTA... 0.00 sec\n",
            "[bwa_index] Construct BWT for the packed sequence...\n",
            "[bwa_index] 0.09 seconds elapse.\n",
            "[bwa_index] Update BWT... 0.00 sec\n",
            "[bwa_index] Pack forward-only FASTA... 0.00 sec\n",
            "[bwa_index] Construct SA from BWT and Occ... 0.04 sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa index MG1655-K12.first400K.fasta\n",
            "[main] Real time: 0.211 sec; CPU: 0.192 sec\n",
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48352, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.27, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.165 CPU sec, 2.859 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48401, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.22, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.193 CPU sec, 2.820 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48399, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.268 CPU sec, 2.907 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48377, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.38, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.183 CPU sec, 2.771 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48371, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.32, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.208 CPU sec, 2.809 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48414, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.36, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.243 CPU sec, 2.900 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48317, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.292 CPU sec, 2.832 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48419, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.34, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.304 CPU sec, 2.861 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48404, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.40, 10.07)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.265 CPU sec, 2.810 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48388, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.36, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.246 CPU sec, 2.906 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48318, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.18)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.196 CPU sec, 2.861 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48338, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.30, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.186 CPU sec, 2.783 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48261, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.35, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.271 CPU sec, 2.856 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48332, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.35, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.294 CPU sec, 2.934 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48379, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.31, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.267 CPU sec, 2.792 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48396, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.29, 10.11)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.298 CPU sec, 2.838 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48371, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.45, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.157 CPU sec, 2.770 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48421, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.38, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.196 CPU sec, 2.865 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48381, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
            "[M::mem_pestat] mean and std.dev: (214.26, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.219 CPU sec, 2.780 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48426, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.32, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.275 CPU sec, 2.838 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48336, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.261 CPU sec, 2.806 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48423, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.11)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.274 CPU sec, 2.899 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48404, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.32, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.180 CPU sec, 2.778 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48382, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.31, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.183 CPU sec, 2.767 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48355, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.31, 10.12)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.153 CPU sec, 2.786 real sec\n",
            "[M::process] read 100000 sequences (10000000 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48327, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.25, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 3.209 CPU sec, 2.906 real sec\n",
            "[M::process] read 16052 sequences (1605200 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48388, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.28, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 100000 reads in 2.903 CPU sec, 2.700 real sec\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 7765, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.28, 10.29)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 16052 reads in 0.641 CPU sec, 0.489 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem MG1655-K12.first400K.fasta corrected_spades/corrected/ecoli_400K_err_1.fastq.00.0_0.cor.fastq.gz corrected_spades/corrected/ecoli_400K_err_2.fastq.00.0_0.cor.fastq.gz\n",
            "[main] Real time: 78.653 sec; CPU: 87.825 sec\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c3ujd_DW_2gE",
        "outputId": "6bc2f6cb-a9b6-45a4-e238-145f19eee1b6"
      },
      "source": [
        "!samtools flagstat alignment_spades_sorted.bam"
      ],
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "2716052 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "0 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "2715942 + 0 mapped (100.00% : N/A)\n",
            "2716052 + 0 paired in sequencing\n",
            "1358026 + 0 read1\n",
            "1358026 + 0 read2\n",
            "2714584 + 0 properly paired (99.95% : N/A)\n",
            "2715874 + 0 with itself and mate mapped\n",
            "68 + 0 singletons (0.00% : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oJbXiHnzAeEQ"
      },
      "source": [
        "# Aligning Trimmomatic results"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "i3q_8QMzHN90",
        "outputId": "e528552a-025d-4a85-a796-09142ec4eb30"
      },
      "source": [
        "!TrimmomaticPE -phred33 ecoli_400K_err_1.fastq.gz ecoli_400K_err_2.fastq.gz -baseout corrected_trim LEADING:20 TRAILING:20  SLIDINGWINDOW:10:20 MINLEN:20"
      ],
      "execution_count": 42,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TrimmomaticPE: Started with arguments:\n",
            " -phred33 ecoli_400K_err_1.fastq.gz ecoli_400K_err_2.fastq.gz -baseout corrected_trim LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20\n",
            "Multiple cores found: Using 2 threads\n",
            "Using templated Output files: corrected_trim_1P corrected_trim_1U corrected_trim_2P corrected_trim_2U\n",
            "Input Read Pairs: 1381602 Both Surviving: 1344536 (97.32%) Forward Only Surviving: 23598 (1.71%) Reverse Only Surviving: 10815 (0.78%) Dropped: 2653 (0.19%)\n",
            "TrimmomaticPE: Completed successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6W1lC7SjHsSE",
        "outputId": "c587d895-b663-433b-f707-1e781864625a"
      },
      "source": [
        "!bwa mem MG1655-K12.first400K.fasta corrected_trim_1P corrected_trim_2P > alignment_trim.sam\n",
        "!samtools view -S -b alignment_trim.sam > alignment_trim.bam\n",
        "!samtools sort alignment_trim.bam > alignment_trim_sorted.bam\n",
        "!samtools index alignment_trim_sorted.bam"
      ],
      "execution_count": 43,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 110200 sequences (10000110 bp)...\n",
            "[M::process] read 110814 sequences (10000040 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53018, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.25, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 110200 reads in 3.044 CPU sec, 3.000 real sec\n",
            "[M::process] read 111006 sequences (10000050 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53373, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.23, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 110814 reads in 3.118 CPU sec, 2.945 real sec\n",
            "[M::process] read 111644 sequences (10000082 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53399, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.34, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 111006 reads in 3.120 CPU sec, 2.942 real sec\n",
            "[M::process] read 112030 sequences (10000010 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53764, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.34, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 111644 reads in 3.113 CPU sec, 2.930 real sec\n",
            "[M::process] read 113494 sequences (10000135 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53899, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.31, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 112030 reads in 3.153 CPU sec, 3.018 real sec\n",
            "[M::process] read 112048 sequences (10000183 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54584, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
            "[M::mem_pestat] mean and std.dev: (214.25, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 113494 reads in 3.188 CPU sec, 3.022 real sec\n",
            "[M::process] read 112288 sequences (10000030 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53846, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.33, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 112048 reads in 3.204 CPU sec, 3.060 real sec\n",
            "[M::process] read 113242 sequences (10000098 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54064, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.37, 10.09)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 112288 reads in 3.221 CPU sec, 2.985 real sec\n",
            "[M::process] read 113776 sequences (10000130 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54423, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.32, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 113242 reads in 3.270 CPU sec, 3.151 real sec\n",
            "[M::process] read 114870 sequences (10000069 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54646, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 113776 reads in 3.267 CPU sec, 3.062 real sec\n",
            "[M::process] read 114452 sequences (10000152 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 55107, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.32, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 114870 reads in 3.273 CPU sec, 3.053 real sec\n",
            "[M::process] read 110288 sequences (10000060 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54846, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.37, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 114452 reads in 3.264 CPU sec, 3.054 real sec\n",
            "[M::process] read 109970 sequences (10000159 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53098, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.18)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 110288 reads in 3.202 CPU sec, 3.038 real sec\n",
            "[M::process] read 108832 sequences (10000054 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52915, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.30, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 109970 reads in 3.144 CPU sec, 2.950 real sec\n",
            "[M::process] read 108132 sequences (10000172 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52350, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.43, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 108832 reads in 3.131 CPU sec, 2.925 real sec\n",
            "[M::process] read 107444 sequences (10000045 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52116, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.36, 10.13)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 108132 reads in 3.198 CPU sec, 3.105 real sec\n",
            "[M::process] read 107442 sequences (10000155 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51765, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.24, 10.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 107444 reads in 3.295 CPU sec, 3.106 real sec\n",
            "[M::process] read 107624 sequences (10000069 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51802, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.29, 10.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 107442 reads in 3.146 CPU sec, 2.945 real sec\n",
            "[M::process] read 107672 sequences (10000147 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51724, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
            "[M::mem_pestat] mean and std.dev: (214.23, 10.10)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 107624 reads in 3.173 CPU sec, 2.987 real sec\n",
            "[M::process] read 107400 sequences (10000070 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51917, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.23, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 107672 reads in 3.157 CPU sec, 2.950 real sec\n",
            "[M::process] read 106862 sequences (10000171 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51693, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.32, 10.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 107400 reads in 3.218 CPU sec, 3.087 real sec\n",
            "[M::process] read 106892 sequences (10000079 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51492, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.29, 10.18)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 106862 reads in 3.080 CPU sec, 2.881 real sec\n",
            "[M::process] read 106614 sequences (10000185 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51426, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.22, 10.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 106892 reads in 3.148 CPU sec, 2.960 real sec\n",
            "[M::process] read 44036 sequences (4155865 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51260, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.31, 10.11)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 106614 reads in 3.049 CPU sec, 2.883 real sec\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 21248, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
            "[M::mem_pestat] mean and std.dev: (214.20, 10.24)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 44036 reads in 1.372 CPU sec, 1.305 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem MG1655-K12.first400K.fasta corrected_trim_1P corrected_trim_2P\n",
            "[main] Real time: 73.773 sec; CPU: 77.648 sec\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hXzoCdv9IE3s",
        "outputId": "0cafbc07-6ff7-4b49-9682-ca6015d1445b"
      },
      "source": [
        "!samtools flagstat alignment_trim_sorted.bam"
      ],
      "execution_count": 44,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "2689072 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "0 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "2687824 + 0 mapped (99.95% : N/A)\n",
            "2689072 + 0 paired in sequencing\n",
            "1344536 + 0 read1\n",
            "1344536 + 0 read2\n",
            "2685506 + 0 properly paired (99.87% : N/A)\n",
            "2686782 + 0 with itself and mate mapped\n",
            "1042 + 0 singletons (0.04% : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UZZSfljvAhUR"
      },
      "source": [
        "# Counting errors"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lBMCFDEEK5P8",
        "outputId": "8c8f5d0d-eebe-4cf9-8f43-dc7cec375860"
      },
      "source": [
        "!pip install pysam\n",
        "!pip install tqdm\n",
        "import pysam\n",
        "from tqdm.notebook import tqdm\n",
        "import pandas as pd"
      ],
      "execution_count": 45,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: pysam in /usr/local/lib/python3.7/dist-packages (0.16.0.1)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.7/dist-packages (4.41.1)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sgN9pho0TduT"
      },
      "source": [
        "Для каждого рида из исправленного и обычного выравнивания будем строить словари, где ключ -- имя рида, значение -- сам рид.\n",
        "После этого проходимся по ним, ищем риды, у которых ID совпадают, для них уже смотрим соответствие (также нахоим этот кусок в референсе).\n",
        "В сам ответ записываем просто сравнивая позицию в референсе с позицией в старом и новом выравниваниях. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "yezNifAyxVSk"
      },
      "source": [
        "def count_errors(samfile_raw: pysam.AlignmentFile, samfile_corrected: pysam.AlignmentFile):\n",
        "  errors = [[0 for _ in range(3)] for _ in range(2)]\n",
        "  raw_dict = {}\n",
        "  for read in samfile_raw.fetch():\n",
        "    if not read.is_unmapped:\n",
        "      raw_dict[read.query_name] = read\n",
        "  print(\"created raw dictionary\")\n",
        "\n",
        "  corrected_dict = {}\n",
        "  for read in samfile_corrected.fetch():\n",
        "    if not read.is_unmapped:\n",
        "      corrected_dict[read.query_name] = read\n",
        "  print(\"created corrected dictionary\")\n",
        "\n",
        "  for read_name, read_raw in tqdm(raw_dict.items()):\n",
        "    dict_raw = {ref_i: None if read_i is None else read_raw.query_sequence[read_i] \\\n",
        "                                   for read_i, ref_i in read_raw.get_aligned_pairs()}\n",
        "    if read_name in corrected_dict:\n",
        "      read_corr = corrected_dict[read_name]\n",
        "      dict_corr = {ref_i: None if read_i is None else read_corr.query_sequence[read_i] \\\n",
        "                                   for read_i, ref_i in read_corr.get_aligned_pairs()}\n",
        "      dict_ref = {}\n",
        "      for read_i, ref_i, ref_l in read_raw.get_aligned_pairs(with_seq = True):\n",
        "        dict_ref[ref_i] = ref_l\n",
        "      for index in dict_raw:\n",
        "        raw_value = dict_raw[index]\n",
        "        cor_value = dict_corr[index] if index in dict_corr else None\n",
        "        ref_value = dict_ref[index] if index in dict_ref else None\n",
        "        if cor_value == 'N':\n",
        "          errors[int(ref_value == raw_value)][2] += 1\n",
        "        else:\n",
        "          errors[int(ref_value == raw_value)][int(ref_value == cor_value)] += 1\n",
        "  return pd.DataFrame(errors)"
      ],
      "execution_count": 46,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 166,
          "referenced_widgets": [
            "67162be1b6d24da2896c004761cfd9df",
            "d95799b233f14888b95194b5050a65c7",
            "8732aaf4719b427bb5df4b605b94e4d2",
            "9a505db5255541e89adb6a9fe45ba33a",
            "49cce6045ff4464abb7740d76cf4978c",
            "b85281b697e24408ae6959659ddb23c3",
            "397768dd75d7491a8d26f4bbfc81814c",
            "5ebda31e1ef8457da4ed944a0d2bfdf9"
          ]
        },
        "id": "MB2Emko_K3qk",
        "outputId": "645eff68-d931-4c41-9728-67c6635af59a"
      },
      "source": [
        "print('PROCESSING THE TRIMMOMATIC RESULTS')\n",
        "samfile_trim = pysam.AlignmentFile(\"alignment_trim_sorted.bam\", \"rb\")\n",
        "samfile = pysam.AlignmentFile(\"alignment_sorted.bam\", \"rb\")\n",
        "print(count_errors(samfile_raw=samfile, samfile_corrected=samfile_trim))"
      ],
      "execution_count": 47,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "PROCESSING THE TRIMMOMATIC RESULTS\n",
            "created raw dictionary\n",
            "created corrected dictionary\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "67162be1b6d24da2896c004761cfd9df",
              "version_minor": 0,
              "version_major": 2
            },
            "text/plain": [
              "HBox(children=(FloatProgress(value=0.0, max=1381551.0), HTML(value='')))"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "\n",
            "         0          1      2\n",
            "0  1006661     134359  68596\n",
            "1  9539618  121291214      1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 166,
          "referenced_widgets": [
            "4c145b72aa274c14aab271fd7bec0e5c",
            "4e4dc3b55948491b91df547fb41269ed",
            "ba4dfd53d44f40f3854a6efa04e93515",
            "966f7a5c714e409bb9f5a518e24a486c",
            "d7536ba565414217b20ca64b122f8f7b",
            "3ff6df99cf824775b9b6ba3c614f41ea",
            "5dd9329e4a3f45aa9c7998b85a88b884",
            "87bfba00b1b94ff2ae133a2f2bbb23f3"
          ]
        },
        "id": "e9aO2QcHLZl2",
        "outputId": "f76702a5-d67d-41c7-ebb7-5ff853791dad"
      },
      "source": [
        "print('PROCESSING THE SPADES RESULTS')\n",
        "samfile_trim = pysam.AlignmentFile(\"alignment_spades_sorted.bam\", \"rb\")\n",
        "samfile = pysam.AlignmentFile(\"alignment_sorted.bam\", \"rb\")\n",
        "print(count_errors(samfile_raw=samfile, samfile_corrected=samfile_trim))"
      ],
      "execution_count": 48,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "PROCESSING THE SPADES RESULTS\n",
            "created raw dictionary\n",
            "created corrected dictionary\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "4c145b72aa274c14aab271fd7bec0e5c",
              "version_minor": 0,
              "version_major": 2
            },
            "text/plain": [
              "HBox(children=(FloatProgress(value=0.0, max=1381551.0), HTML(value='')))"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "\n",
            "         0          1       2\n",
            "0  1087323       5357  144007\n",
            "1  7816149  124124489  189747\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Deqnw3tiTci0"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}